pkg/controller/machine.go

```diff
@@ -614,6 +651,23 @@ func (c *controller) machineDelete(machine *v1alpha1.Machine, driver driver.Driv
 			}
 
 			err = driver.Delete(machineID)
+		} else {
+			klog.V(2).Infof("Missing ProviderID when deleting the machine object. %s might have been deleted too soon.", machine.Name)
+			// As MachineID is missing, we should check once if actual VM was created but MachineID was not updated on machine-object.
+			// We list VMs and check if any one them map with the given machine-object.
+			VMList, _ := driver.GetVMs("")
+			for providerID, machineName := range VMList {
+				if machineName == machine.Name {
+					klog.V(2).Infof("Deleting the VM %s backing the machine-object %s.", providerID, machine.Name)
+					err = driver.Delete(providerID)
+					if err != nil {
+						klog.Errorf("Error deleting the VM %s backing the machine-object %s due to error %v", providerID, machine.Name, err)
+						// Not returning error so that status on the machine object can be updated in the next step if errored.
```

Comment:
I didn't follow this. Won't the machine object be deleted in the next step and this orphan VM continues to be there? (unless the orphan VM controller acts on it). And incase of shoot deletion this might still lead to situations where VM persists once the machine object is deleted?

Reply:
If `err` is set here by `driver.Delete()`, then the check in the next block `err != nil`should return the same `err` after setting it on the machine-status.

Reply:
Okay, then it's alright. I didn't understand the comment earlier.